### PR TITLE
Skeletal socket transport

### PR DESCRIPTION
**Pull request by jakewins**
Does not yet support the actual message protocol, this is the module skeleton
and a proposal for version negotiation. 

Choosing a version needs to be highly
flexible for future protocol changes. This PR lets the client propose up to four
supported protocol versions, in order of preference, and the server then chooses
the first one it supports (or tells the client none will work):

```
Client: <connect>
Client: 0 0 0 2  0 0 0 1  0 0 0 0  0 0 0 0  // I support version '2' and '1', and prefer '2'
Server: 0 0 0 1 // Version '1' will have to do
<begin versioned messaging>
```

This is done outside the scope of PackStream (allowing it to change cross
versions), the messaging protocol and the (yet-to-be-created) framing protocol.

The number of proposed versions is bounded to 4 (if the client wants to try
more it can simply send a new version proposal, having it be fixed-width lowers
complexity), each version is a 4-byte unsigned integer, bounding the number of
protocol versions we can have to 4 294 967 295, which is a very large number and
should be ample time for Elon Musk to sort all this 'computer' stuff out for us.

See SocketSessionIT for examples of negotiating protocol version.

### CODE REVIEW

**Comment by jakewins** [comment]
@zhenlineo that is on purpose, yes. Netty 4 adds buffer pooling, before that it allocates a massive amount of buffers as traffic arrives and departs. Netty 4 is built in a way where it can co-exist with Netty 3, so I figured this would be a good first module to move to Netty 4, and we can shift Com and Cluster over one at a time afterwards. 

@lutovich +1, I'll modify this to have all the versions pulled out of the parent

File: community/ndp/transport-socket/src/main/java/org/neo4j/ndp/transport/socket/ChunkedInput.java [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+        }
+        return this;
+    }
+
+    public void addChunk( ByteBuf chunk )
+    {
+        if(chunk.readableBytes() > 0)
+        {
+            byte[] dst = new byte[chunk.readableBytes()];
+            chunk.getBytes( 0, dst );
```
Comment (zhenlineo) [inline community/ndp/transport-socket/src/main/java/org/neo4j/ndp/transport/socket/ChunkedInput.java]: I am confused, why we need the variable `dst`?

**Comment by zhenlineo** [comment]
Also your code format freestyle is a bit different from ours. Please formatting :)

File: community/ndp/transport-socket/src/main/java/org/neo4j/ndp/transport/socket/ChunkedInput.java (hunk 1 above)
Comment (jakewins) [inline community/ndp/transport-socket/src/main/java/org/neo4j/ndp/transport/socket/ChunkedInput.java]: Lol, remnants from an earlier approach, good catch.

File: community/ndp/messaging-v1/src/test/java/org/neo4j/ndp/messaging/v1/util/MessageMatchers.java [hunk 2]
```diff
@@ -135,69 +131,46 @@ public void describeTo( Description description )
...
-            public Object content()
-            {
-                return payload;
-            }
-
-            @Override
-            public MediaType contentType()
-            {
-                return MediaType.valueOf( PackStreamMessageFormatV1.CONTENT_TYPE );
+                description.appendText( "ITEM " );
```
Comment (technige) [inline community/ndp/messaging-v1/src/test/java/org/neo4j/ndp/messaging/v1/util/MessageMatchers.java]: Should this be `RECORD`?

**Pull request merged**